pkg/providers/vault/client.go

```diff
@@ -90,13 +91,63 @@ func (c Client) getAllKeys(path string) (map[string]string, error) {
 		req.Header.Set(NamespaceHeader, c.HttpConfig.Namespace)
 	}
 
-	resp, err := c.HttpCaller.Do(req)
-	if err != nil {
-		return nil, err
-	}
+	var resp *http.Response
+	addRetryAttempts := c.HttpConfig.AdditionalRetryAttempts
+	switch {
+	case addRetryAttempts < 0:
+		return nil, pkg.NewErrSecretStore(fmt.Sprintf("invalid retry attempts setting %d", addRetryAttempts))
+		// the number of retries is infinite
+		// TODO: is this useful? leaving this here in case it's useful eventually
+		// needs tests though...
+		// for {
+		// 	resp, err = c.HttpCaller.Do(req)
+		// 	if err != nil {
+		// 		return nil, err
+		// 	}
+
+		// 	// don't distribute the not here to keep the logic simple to
+		// 	// understand
+		// 	if !(resp.StatusCode < 200 || resp.StatusCode > 299) {
+		// 		break
+		// 	}
+
+		// 	time.Sleep(c.HttpConfig.RetryWaitPeriod)
+		// }
+	case addRetryAttempts == 0:
+		// no retries
+		resp, err = c.HttpCaller.Do(req)
+		if err != nil {
+			return nil, err
+		}
 
-	if resp.StatusCode < 200 || resp.StatusCode > 299 {
-		return nil, pkg.NewErrSecretStore(fmt.Sprintf("Received a '%d' response from the secret store", resp.StatusCode))
+		if resp.StatusCode < 200 || resp.StatusCode > 299 {
+			return nil, pkg.NewErrSecretStore(fmt.Sprintf("Received a '%d' response from the secret store", resp.StatusCode))
+		}
+
+	case addRetryAttempts > 0:
+		// do some retries
+		// note the limit is 1 + additional retry attempts, cause we always need
+		// to do 1
+		for tryNum := 0; tryNum < 1+addRetryAttempts; tryNum++ {
```

Comment:
I think the retry logic should be moved to the `GetSecrets` method:

https://github.com/edgexfoundry/go-mod-secrets/blob/65a79df9754eb7cfd29c29606b68a90d53ede380/pkg/providers/vault/client.go#L49

During testing I ran into a situation where there were some secrets in Vault, but not all of them. This resulted in the `getAllKeys` successfully returning and the `GetSecrets` method would then return an error, bypassing the retry logic. By moving it to `GetSecrets` we get the retry logic to work with any failing operation while retrieving keys.

Reply:
okay, I can move this to GetSecrets, when I initially worked on this I wanted to be able to only retry on http errors and not on other catastrophic errors, I will submit a PR a bit later today doing this